experiment/run_experiment.py

```diff
@@ -242,15 +266,114 @@ def copy_resources_to_bucket(config_dir: str, config: Dict):
     gsutil.rsync(config_dir, destination)
 
 
-class Dispatcher:
-    """Class representing the dispatcher instance."""
+class BaseDispatcher:
+    """Class representing the dispatcher."""
+
+    def __init__(self, config: Dict):
+        self.config = config
+        self.instance_name = experiment_utils.get_dispatcher_instance_name(
+            config['experiment'])
+        self.process = None
+
+    def create_async(self):
+        """Creates the dispatcher asynchronously."""
+        raise NotImplementedError
+
+    def start(self):
+        """Start the experiment on the dispatcher."""
+        raise NotImplementedError
+
+
+class LocalDispatcher:
+    """Class representing the local dispatcher."""
 
     def __init__(self, config: Dict):
         self.config = config
         self.instance_name = experiment_utils.get_dispatcher_instance_name(
             config['experiment'])
         self.process = None
 
+    def create_async(self):
+        """Noop in local experiments."""
+
+    def start(self):
+        """Start the experiment on the dispatcher."""
+        shared_volume_dir = os.path.abspath('shared-volume')
+        if not os.path.exists(shared_volume_dir):
+            os.mkdir(shared_volume_dir)
+        shared_volume_volume_arg = '{0}:{0}'.format(shared_volume_dir)
+        shared_volume_env_arg = 'SHARED_VOLUME={}'.format(shared_volume_dir)
+        sql_database_arg = 'SQL_DATABASE_URL=sqlite:///{}'.format(
+            os.path.join(shared_volume_dir, 'local.db'))
+
+        home = os.environ['HOME']
+        host_gcloud_config_arg = (
+            'HOST_GCLOUD_CONFIG={home}/{gcloud_config_dir}'.format(
+                home=home, gcloud_config_dir='.config/gcloud'))
+
+        base_docker_tag = experiment_utils.get_base_docker_tag(
+            self.config['cloud_project'])
+        set_instance_name_arg = 'INSTANCE_NAME={instance_name}'.format(
+            instance_name=self.instance_name)
+        set_experiment_arg = 'EXPERIMENT={experiment}'.format(
+            experiment=self.config['experiment'])
+        set_cloud_project_arg = 'CLOUD_PROJECT={cloud_project}'.format(
+            cloud_project=self.config['cloud_project'])
+        set_cloud_experiment_bucket_arg = (
+            'CLOUD_EXPERIMENT_BUCKET={cloud_experiment_bucket}'.format(
+                cloud_experiment_bucket=self.config['cloud_experiment_bucket']))
+        docker_image_url = '{base_docker_tag}/dispatcher-image'.format(
+            base_docker_tag=base_docker_tag)
+        volume_arg = '{home}/.config/gcloud:/root/.config/gcloud'.format(
```

Comment:
Is gcloud credentials for gcs. this is still pretty ugly, we need to remove gcs dependency sometime.

Reply:
Yep. I agree about GCS dependancy.

Reply:
would it be more work to start with abstracting away the file storage system first? by that i mean adding a generic file interface in front of the current GCS client implementation, and then add a local file system implementation of the same interface. there are also libraries for this, that support GCS, Local FS, and other cloud storage systems as well (e.g., https://pypi.org/project/cloudstorage/).

Reply:
It's easier for me to assume we can use GCS at first, and then remove this dependancy than doing everything at once. I like that library and we can try using it in a later iteration.

Reply:
sounds good